src/metabase/api/database.clj

```diff
@@ -713,6 +724,71 @@
              (filter field-perm-check (-> (database/pk-fields {:id id})
                                           (t2/hydrate :table))))))
 
+;;; ----------------------------------------- GET /api/database/:id/table/:fqn ---------------------------------------
+
+(defn- check-unique [tables]
+  (when (seq tables)
+    (let [unique-combos (distinct (map (juxt :schema :name) tables))]
+      (when (> (count unique-combos) 1)
+        (throw (ex-info "Ambiguous table identifier"
+                        {:status-code       300
+                         :potential-matches (for [[schema table-name] unique-combos]
+                                              (if schema
+                                                (str schema "." table-name)
+                                                table-name))})))
+      (first tables))))
+
+(defn- resolve-table [db-id table-identifier]
+  (let [[schema table-name] (let [parts (str/split table-identifier #"\.")]
+                              (case (count parts)
+                                1 [nil (first parts)]
+                                2 parts
+                                nil))]
+    (api/check table-name 400 "Invalid table identifier")
+    (api/check-404
+     (or (check-unique (t2/select :model/Table
+                                  :db_id db-id
+                                  :name table-name
+                                  (cond-> {}
+                                    schema (assoc :where {:schema schema}))))
+         (check-unique (t2/select :model/Table
+                                  :db_id db-id
+                                  :%lower.name (u/lower-case-en table-name)
+                                  (cond-> {}
+                                    schema (assoc :where {:%lower.schema (u/lower-case-en schema)}))))))))
+
+(api.macros/defendpoint :get "/:db-id/table/:table-identifier/data"
+  "Get the data for the given table"
+  [{:keys [db-id table-identifier]} :- [:map
+                                        [:db-id ms/PositiveInt]
+                                        [:table-identifier ms/NonBlankString]]]
+  (api/read-check (t2/select-one :model/Database :id db-id))
+  (qp.store/with-metadata-provider db-id
+    (let [mp    (qp.store/metadata-provider)
+          table (resolve-table db-id table-identifier)
+          query (-> (lib/query mp (lib.metadata/table mp (:id table)))
+                    (update-in [:middleware :js-int-to-string?] (fnil identity true))
+                    qp/userland-query-with-default-constraints
+                    (update :info merge {:executed-by api/*current-user-id*
+                                         :context     :table-grid
+                                         :card-id     nil}))]
+      ;; TODO Discuss whether we want these analytics
+      #_(events/publish-event! :event/table-read {:object  (t2/select-one :model/Table :id table-id)
```

Comment:
Waiting on confirmation in [Slack](https://metaboat.slack.com/archives/C0641E4PB9B/p1741094197625719), but merging with it commented out like this should be fine for now.